envelope.go

```diff
@@ -42,69 +42,105 @@ var errSpecialEnvelope = errorf(
 type envelope struct {
 	Data  *bytes.Buffer
 	Flags uint8
+
+	offset int
 }
 
 func (e *envelope) IsSet(flag uint8) bool {
 	return e.Flags&flag == flag
 }
 
+// Read implements io.Reader.
+func (e *envelope) Read(data []byte) (readN int, err error) {
+	if e.offset < 5 {
+		prefix := makeEnvelopePrefix(e.Flags, e.Data.Len())
+		readN = copy(data, prefix[e.offset:])
+		e.offset += readN
+		if e.offset < 5 {
+			return readN, nil
+		}
+		data = data[readN:]
+	}
+	n := copy(data, e.Data.Bytes()[e.offset-5:])
+	e.offset += n
+	readN += n
+	if readN == 0 && e.offset == e.Data.Len()+5 {
+		err = io.EOF
+	}
+	return readN, err
+}
+
+// WriteTo implements io.WriterTo.
+func (e *envelope) WriteTo(dst io.Writer) (wroteN int64, err error) {
+	if e.offset < 5 {
+		prefix := makeEnvelopePrefix(e.Flags, e.Data.Len())
+		prefixN, err := dst.Write(prefix[e.offset:])
+		e.offset += prefixN
+		wroteN += int64(prefixN)
+		if e.offset < 5 {
+			return wroteN, err
+		}
+	}
+	n, err := dst.Write(e.Data.Bytes()[e.offset-5:])
+	e.offset += n
+	wroteN += int64(n)
+	return wroteN, err
+}
```

Comment:
For #611 will also need to add `func (e *envelope) Rewind() bool { b.offset = 0}`

Reply:
I think it would be way cleaner if `envelope` were immutable once sent. So that means we'd instead want a `NewReader() io.Reader` method, so _that_ thing can be mutable/rewindable. That way, if any synchronization becomes necessary (as exists in the payload thingie in #611), it can live in that reader impl and not leak into the envelope struct itself.